[Bug]: Stop orders cause a panic after snapshot restore

### Problem encountered

When starting a node from a snapshot, there is a runtime panic when expiring stop orders (nil pointer dereference)

### Observed behaviour

Stack trace:

```
panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x15a13ad]


goroutine 1161 [running]:

code.vegaprotocol.io/vega/core/execution/stoporders.(*Pool).RemoveExpired(0xc00e64f6b0, {0xc01ed26580, 0x2, 0x177e4ec33e5f1779?})

	/home/runner/work/vega/vega/core/execution/stoporders/stop_orders.go:259 +0x1ed

code.vegaprotocol.io/vega/core/execution/future.(*Market).removeExpiredStopOrders(0xc00e662500, {0x3e30228, 0xc01ebb27b0}, 0xc01ebdf9c0?, {0x3dfe560, 0xc01e8e3e00})

	/home/runner/work/vega/vega/core/execution/future/market.go:3463 +0x1bc

code.vegaprotocol.io/vega/core/execution/future.(*Market).OnTick(0xc00e662500, {0x3e30228, 0xc01ebb27b0}, {0x40?, 0xc01ed126e0?, 0x0?})

	/home/runner/work/vega/vega/core/execution/future/market.go:664 +0x4a8

code.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0xc001510900, {0x3e30228, 0xc01ebb27b0}, {0x1?, 0x3e4e848?, 0x0?})

	/home/runner/work/vega/vega/core/execution/engine.go:1078 +0x4eb

code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)

	/home/runner/work/vega/vega/core/vegatime/service.go:104

code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0xc000e4aa00, {0x3e30228, 0xc01ebb27b0}, {0x40?, 0x2e36440?, 0x0?})

	/home/runner/work/vega/vega/core/vegatime/service.go:77 +0x24f

code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc01ebb08e0, 0x20, 0x20}, {{0xb, 0x1}, {0xc01eb488a0, 0x1c}, 0xabb6b9, {0x186e7b79, ...}, ...}, ...})

	/home/runner/work/vega/vega/core/processor/abci.go:802 +0x747

code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc01ebb08e0, 0x20, 0x20}, {{0xb, 0x1}, {0xc01eb488a0, 0x1c}, 0xabb6b9, {0x186e7b79, ...}, ...}, ...})

	/home/runner/work/vega/vega/core/blockchain/abci/abci.go:46 +0x75

code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).BeginBlock(_, {{0xc01ebb08e0, 0x20, 0x20}, {{0xb, 0x1}, {0xc01eb488a0, 0x1c}, 0xabb6b9, {0x186e7b79, ...}, ...}, ...})

	/home/runner/work/vega/vega/cmd/vega/commands/node/app_wrapper.go:64 +0x55

github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {{0xc01ebb08e0, 0x20, 0x20}, {{0xb, 0x1}, {0xc01eb488a0, 0x1c}, 0xabb6b9, {0x186e7b79, ...}, ...}, ...})

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/abci/client/local_client.go:280 +0x118

github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc01ebb08e0, 0x20, 0x20}, {{0xb, 0x1}, {0xc01eb488a0, 0x1c}, 0xabb6b9, {0x186e7b79, ...}, ...}, ...})

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/proxy/app_conn.go:81 +0x55

github.com/tendermint/tendermint/state.execBlockOnProxyApp({0x3e2e5f0?, 0xc00012e9a8}, {0x3e451c0, 0xc000eeb460}, 0xc01eb23e00, {0x3e57da0, 0xc0014fff08}, 0xabb6b8?)

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:307 +0x3dd

github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x1}, {0x358b5e4, 0x7}}, {0xc00040b560, 0x1c}, 0x1, 0xabb6b8, {{0xc017b68520, ...}, ...}, ...}, ...)

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:140 +0x171

github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc0004cfdc0, 0x1)

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/blockchain/v0/reactor.go:400 +0xbda

created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).SwitchToFastSync

	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/blockchain/v0/reactor.go:127 +0xe5

script returned exit code 2
``` 

### Expected behaviour

Other nodes don't panic, so there is a problem with the data in the snapshot and/or the way we restore the data. The node should load the state, and run as normal

### Steps to reproduce

```bash
1. Download the snapshots and chain from: https://jenkins.ops.vega.xyz/job/private/job/Snapshots/job/Fairground/33915/artifact/
2. Build vega
3. Replay from the latest snapshot
4. vega will panic shortly thereafter
```


### Software version

release/v0.72.x (26afd41a2)

### Failing test

_No response_

### Jenkins run

https://jenkins.ops.vega.xyz/blue/organizations/jenkins/private%2FSnapshots%2FFairground/detail/Fairground/33915/pipeline/64

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -126,6 +126,7 @@
 - [9153](https://github.com/vegaprotocol/vega/issues/9153) - `MTM` win transfers can be less than one.
 - [9178](https://github.com/vegaprotocol/vega/issues/9178) - Fix LP amendment panic
 - [9053](https://github.com/vegaprotocol/vega/issues/9053) - Handle settle market events in core positions plug-in.
+- [9189](https://github.com/vegaprotocol/vega/issues/9189) - Fix stop orders snapshots
 
 ## 0.72.1
 

--- a/core/execution/future/market.go
+++ b/core/execution/future/market.go
@@ -3391,6 +3391,20 @@ func (m *Market) removeExpiredStopOrders(
 	toExpire := m.expiringStopOrders.Expire(timestamp)
 	stopOrders := m.stopOrders.RemoveExpired(toExpire)
 
+	//  ensure any OCO orders are also expire
+	toExpireSet := map[string]struct{}{}
+	for _, v := range toExpire {
+		toExpireSet[v] = struct{}{}
+	}
+
+	for _, so := range stopOrders {
+		if _, ok := toExpireSet[so.ID]; !ok {
+			if so.Expiry.Expires() {
+				m.expiringStopOrders.RemoveOrder(so.Expiry.ExpiresAt.UnixNano(), so.ID)
+			}
+		}
+	}
+
 	updatedAt := m.timeService.GetTimeNow()
 
 	if m.as.InAuction() {

--- a/core/types/snapshot_nodes.go
+++ b/core/types/snapshot_nodes.go
@@ -3478,13 +3478,25 @@ func ExecMarketFromProto(em *snapshot.Market) *ExecMarket {
 		StopOrders:                 em.StopOrders,
 		Product:                    em.Product,
 	}
+
 	for _, o := range em.ExpiringOrders {
 		or, _ := OrderFromProto(o)
 		ret.ExpiringOrders = append(ret.ExpiringOrders, or)
 	}
-	for _, o := range em.ExpiringStopOrders {
-		ret.ExpiringStopOrders = append(ret.ExpiringOrders, &Order{ID: o.Id, ExpiresAt: o.ExpiresAt})
+
+	// TODO: remove this on the next minor
+	if em.StopOrders != nil {
+		for _, v := range em.StopOrders.StopOrders {
+			if v.StopOrder.ExpiresAt != nil {
+				ret.ExpiringStopOrders = append(ret.ExpiringStopOrders, &Order{ID: v.StopOrder.Id, ExpiresAt: *v.StopOrder.ExpiresAt})
+			}
+		}
 	}
+
+	// TODO: enable this then
+	// for _, o := range em.ExpiringStopOrders {
+	//     ret.ExpiringStopOrders = append(ret.ExpiringStopOrders, &Order{ID: o.Id, ExpiresAt: o.ExpiresAt})
+	// }
 	return &ret
 }
 
@@ -3527,7 +3539,7 @@ func (e ExecMarket) IntoProto() *snapshot.Market {
 		ret.ExpiringOrders = append(ret.ExpiringOrders, o.IntoProto())
 	}
 	for _, o := range e.ExpiringStopOrders {
-		ret.ExpiringStopOrders = append(ret.ExpiringOrders, &vega.Order{Id: o.ID, ExpiresAt: o.ExpiresAt})
+		ret.ExpiringStopOrders = append(ret.ExpiringStopOrders, &vega.Order{Id: o.ID, ExpiresAt: o.ExpiresAt})
 	}
 	return &ret
 }

